language/changes/6-phantom-type-params.md

```diff
@@ -0,0 +1,103 @@
+ability `key`, which required all arguments to `CoinType` to be declared with `store`. Specifically,
+the types `XUS::XUS` and `XDX::XDX` were forced to have a `store` annotation even though they were
+never stored in the database! More over the extra `store` requirement polluted the entire Diem
+Framework and many functions had to specify an extra `store` constraints on their type parameters.
+With the new feature, `Diem::Diem<CoinType>` now declares `CoinType` to be phantom which avoids the
+unnecessary `store` annotations.
+
+## Description
+
+### Declaration
+
+In a struct definition a type parameter can be declared as phantom by adding the `phantom` keyword
+before its declaration. If a type parameter is declared as phantom we say it is a phantom type
+parameter. When defining a struct, Move's type checker ensures that every phantom type parameter is
+either not used inside the struct definition or it is only used as an argument to a phantom type
+parameter.
+
+More formally, if a type is used as an argument to a parameter declared as phantom we say the type
+appears in phantom position. With this definition in place, the rule for the correct use of phantom
+parameters can be specified as follows: **A phantom type parameter can only appear in phantom
+position**.
+
+The following two examples show valid uses of phantom parameters. In the first one, the parameter
+`T1` is not used at all inside the struct definition. In the second one, the parameter `T1` is only
+used as an argument to a parameter also declared as phantom.
+
+```
+struct S1<phantom T1, T2> { f: u64 }
+                  ^^
+                  Ok: T1 does not appear inside the struct definition
+
+
+struct S2<phantom T1, T2> { f: S1<T1, T2> }
+                                  ^^
+                                  Ok: T1 appears in phantom position
+```
+
+The following code shows examples of violations of the rule:
+
+```
+
+struct S1<phantom T> { f: T }
+                          ^
+                          Error: Not a phantom position
+
+struct S2<T> { f: T }
+
+struct S3<phantom T> { f: S2<T> }
+                             ^
+                             Error: Not a phantom position
+```
+
+### Instantiation
+
+When instantiating a struct, the arguments to phantom parameters are excluded when deriving the
+struct abilities. For example, consider the following code:
+
+```
+struct S<T1, phantom T2> has copy { f: T1 }
+struct NoCopy {}
+struct HasCopy has copy {}
+```
+
+Consider now the type `S<HasCopy, NoCopy>`. Since `S` is defined with `copy` and all non-phantom
+arguments have copy then `S<HasCopy, NoCopy>` also has copy.
+
+### Phantom Type Parameters with Ability Constraints
+
+Ability constraints and phantom type parameters are orthogonal features in the sense that phantom
+parameters can be declared with ability constraints. And any type parameter must satisfy that
+constraint, even thought it is phantom. For example, the following definition is perfectly valid:
```

Comment:
A few nits on this sentence:
- typo: "thought" should be "though
- It's not a complete sentence. At a minimum, you need a comma after the initial "and", but probably there's a better way.
- It is not clear what "that constraint" refers to. The preceding sentence refers to "ability constraints" in general but says nothing about a specific constraint.
- It is also not clear what "it" refers to. Presumably "any type parameter", but it's confusing because it looks like it could be "that constraint" (which makes no sense).